### DISCUSSION TITLE
Can NequIP use ZBL to prevent atoms from being too close together?

### DISCUSSION THREAD

**Discussion by ChengQIAN93** [mir-group/nequip#475] [Q&A] [answered]
Hi, everyone, 

I've been using NequIP to train my system. I added structures with short interatomic distances (e.g., 1.0 to 1.5 angstroms) to the training set to prevent atoms from being too close together. 

However, some simulations show that the interatomic distance is still too close, which makes a very low potential energy.

My quesiton is can NequIP use ZBL or another method to make atoms repel each other when they're too close?

Thank you.

**Comment by cw-tan** [comment, marked as answer]
Hi @ChengQIAN93 ,

Thank you for your interest in NequIP. It is possible to include the ZBL term in a NequIP GNN model as shown in this tutorial [config file](https://github.com/mir-group/nequip/blob/main/configs/minimal_pair.yaml). The nuclear repulsion effect of ZBL should provide a strong prior to make atoms repel each other when they get too close, but it could be possible that either the repulsion is too short-range (it would also depend on the polynomial cutoff's `p` hyperparameter as that dictates how the ZBL term decays to zero at the radial cutoff), or that the NequIP model was initialized in a way that counteracts the ZBL near atom cores (though this effect is unlikely to be too significant). We have seen empirically that ZBL helps make MD more stable by preventing atoms from getting too close to each other, so it's definitely worth a shot. Adding structures with short interatomic distances would be helpful too, and a combination of both ZBL and including such frames would likely be useful.

**Reply by ChengQIAN93** [reply]
Thanks for your kindly reply.  I tried your suggestion of using ‘pair_style’ and it worked well. I have one more question. If I need to use ‘zbl’, do I need to include some structures with short distances for the fitting of zbl parameters?

**Reply by cw-tan** [reply]
There's no fitting of ZBL parameters, the parameters are deterministic from the atomic species. See code [here](https://github.com/mir-group/nequip/blob/1e150cdc8614e640116d11e085d8e5e45b21e94d/nequip/nn/pair_potential.py#L317). In general though, it would be good to include some structures with short distances to condition the model on them. While the ZBL term is deterministic, I don't have a good sense on how the untrained NequIP GNN model will behave for the short distance regime and how it would evolve during training if it's never seen atomic environments with short atomic distances -- adding those structures to your training should make the overall model more robust (learned NequIP GNN model + ZBL potential).

**Reply by rushikd27** [reply]
Hi, 

I wanted to know if there is a way to extract these ZBL energies. Another question I had was about a high, constant cutoff, which might induce spurious ZBL energies. It would be great if you could help with this.

Thank you,
Rushik

**Reply by cw-tan** [reply]
Because of the way the model is constructed, I can think of two ways (but not that straightforward unfortunately)

- start with the original model with ZBL, and make a copy where ZBL is popped out of the model. The different in the predictions is the ZBL contribution
- instantiate a new ZBL model, making sure that what it computes exactly aligns with the ZBL module in the model (e.g. polynomial cutoff function used, etc)

By high, constant cutoff, do you mean using the same ~6 A cutoff radius as the rest of the model? The cutoff envelope function may help, and one could choose say p=2 for the polynomial cutoff function which makes the attenuation a lot more rapid.

**Reply by rushikd27** [reply]
Thank you for the quick response! I will try reducing the degree of the overlap function.

In the first approach, though, even if we keep everything the same, training the model with ZBL might tweak it internally, so the difference cannot be certainly attributed to ZBL alone. Or am I misinterpreting how the ZBL energies and forces are incorporated during training?

**Reply by cw-tan** [reply]
I was thinking more about a model that you've already trained with ZBL on, i.e. if you just want to analyze the decomposition of energies between the model and ZBL.

**Reply by rushikd27** [reply]
Ahh, I did that and found that the ZBL is being applied to almost all the training structures. It is an allegro model, though. Does that make a difference in your experience?
